### ISSUE TITLE
ROLLBACK Transaction with Code / SQL Error

### ISSUE THREAD

**Issue by BallisticPain** [cakephp/phinx#896]
I was under the impression that using transactions for migrations was for when either you had code or SQL errors it could rollback the migration so you could simply fix and migrate up. From what I saw in the code that isn't the case.

We run into a lot of (usually code errors) and then have to manually undo the changes that were made successfully. I originally just thought our version of MySQL didn't have transactions, but we're on 5.5. So that's when I looked in the code.

Would anyone be opposed to that? I believe adding a `try...catch` would do the trick quite nicely. Not sure that would catch any SQL errors that are generated, but it certainly would catch the PHP errors.

If this is of interest, I would be happy to write the PR for this.

**Comment by rquadling** [comment]
As far as I understand it (not an expert in all DBs), schema changes cannot be transacted. So adding several tables and one fails, no rollback possible.

**Comment by BallisticPain** [comment]
I was just main painfully aware of this specifically for MySQL. One thing that South (now in Django) did was it spit out SQL for the `backwards` method... this might be handy... thoughts?

**Comment by rquadling** [comment]
I see what you are saying. Only would be feasible in a `change()` call as they have the reversible calls. You would have your work cut out dealing with determining if a migration contains schema and data changes.

As I don't have any sort of stats on how often a user would require such a feature, I don't know if it would be worth spending any time on it.

**Comment by BallisticPain** [comment]
Postgres does allow schema transactions...

The South migration library simply took the SQL output of the `backwards` method and spit that out to the command line. Below is an output I ran into the past couple of days...

```
! Error found during real run of migration! Aborting.

 ! Since you have a database that does not support running
 ! schema-altering statements in transactions, we have had 
 ! to leave it in an interim state between migrations.

! You *might* be able to recover with:   = ALTER TABLE `plans` ; []
   = ALTER TABLE `plans` ALTER COLUMN `position_desired_id` DROP DEFAULT; []
   = ALTER TABLE `plans` MODIFY `position_desired_id` integer NOT NULL;; []
   = ALTER TABLE `plans` ADD CONSTRAINT `position_desired_id_refs_job_id_77e1f9ad00e9d6d` FOREIGN KEY (`position_desired_id`) REFERENCES `positions` (`job_id`); []

 ! The South developers regret this has happened, and would
 ! like to gently persuade you to consider a slightly
 ! easier-to-deal-with DBMS.

Traceback (most recent call last):
```

As you can see their "might be able to recover with" doesn't even contain the full needs to recover. Which I'm sure is likely a bug with the version I'm stuck dealing with at the moment...

So what do you think the effort would be to get the SQL of the down/change method to spit out along with the error?

**Comment by rquadling** [comment]
Top of my head / rough ideas / no guarantees implied / YMMV / etc. / etc.

If an `up()`, then sort of running `down()` in a --dry-run sort of mode that grabbed the SQL and reported it.

One issue is the up and down may not be connected. We have many instances where there is no down(). For example, data correction.

For change(), well, the same thing I suppose, pretend to rollback the migration, in dry-run mode, with a way to force it to run, even though the migration actually hasn't.

You are going to get screwed by any logic in the migration (i.e. hasColumn() sort of thing) as these will fail in the down\rollback when you expected them to be OK, so, again, you still may not get any output.

So, step one, dry-run, be able to run through a migration without it having any effect. That has been asked for before.

Once you've got a dry run, you SHOULD be able to migrate/rollback and log all the queries.

You may also have to limit the number of queries. One of my migrations has to update over a million rows with a specific value based upon pre-existing data. So a reverse order iterator was used. A million rows being reported would be unfeasible. You'd probably run out of memory holding them.

But that was in an `up()` with no `down()`, so only the dry-run of the `down()` would be right. For a `change()`, it is possible that all the queries could be determined and output unbuffered.

**Comment by ZaneCEO** [comment]
I found this issue looking for the same thing. Without schema-changes-transaction-support it's definitely not trivial to find a good all-around solution. Meanwhile, can I suggest to implement transaction+commit/rollback for data-only operations? This would be a lifesaver and with some minimal planning user-side we can put all the (% few!) schema-related operations in one migration and all the (% many!) data-related operation in another, splitting the risk.

**Comment by rquadling** [comment]
I sort of half thought that transaction+commit/rollback was already in place, but I can see that it isn't fully.

Considering it is possible to have a migration that uses raw SQL, it is going to be a bit difficult to guarantee the split. I think it may simply be best to leave the issue up to the user.

As an aside, I've got instances where I don't want to use transactions (when updating millions of rows one at a time as changing ALL the rows in one go locks the table for far too long). I want to be able to disable migration wide transactioning and handle it myself.

I currently do that by calling
```php
        // Disable the transaction locking.
        $this->getAdapter()->commitTransaction();
```

So having a mechanism to disable transactions and/or automatic rollback upon failure would be ideal if try/catch/rollback was implemented.
